### ISSUE TITLE
Update to new Charite dataset

### ISSUE THREAD

**Issue by ajratner** [ajratner/dd-genomics#30]
See email from Harendra- the dataset he and Aaron are now using

**Comment by Colossus** [comment]
@ajratner Hey Alex, would you mind if I took on this one? It's probably going to take quite a bit longer than if you did, but I think it would be a learning experience. If it's OK, do you know where to get the dataset from?

**Comment by ajratner** [comment]
@Colossus I mind terribly :)  Sounds great!

From @hguturu : New combined dataset is at `/cluster/data/genomes/human/grch37/medgendb/latest/hpo/phenotype_to_gene.map`

**Comment by Colossus** [comment]
I need to figure out some stuff about this new dataset. A few facts.

Old dataset: 7023 HPO phenotypes, 3185 genes (in canonical form (?)), 286249 lines in file (= associations)
New dataset: 6074 HPO phenotypes, 8268 genes (in Ensemble form), 142509 lines in file (= associations)

**Comment by Colossus** [comment]
OK, you can apparently also download a bunch of data here, which is where the "old" charite supervision data set is being automatically downloaded for: http://compbio.charite.de/hudson/job/hpo.annotations.monthly/lastStableBuild/artifact/annotation/

**Comment by Colossus** [comment]
A huge mail thread went into this, reproduced here:

---

Hey Aaron,

You mentioned that you guys had just put in a new version of Charite with more GP pairs, as well as an additional dataset- do you think you could direct me to these so we can (either now or soon down the road) integrate them as supervision data into DD?

Thanks!
Alex

---

Hey Alex,

You can find the new combined dataset can be found at /cluster/data/genomes/human/grch37/medgendb/latest/hpo/phenotype_to_gene.map.

Might be worthwhile to keep these as a test set for later on in case you actually are doing supervised learning (since they are objectively independent of your training set).

Harendra

---

Hi Harendra and Aaron,

I'm just trying to integrate this new Charite dataset into our supervision data. A few questions about this:

Currently, the Charite data set is just being downloaded from http://compbio.charite.de/hudson/job/hpo.annotations.monthly/lastStableBuild/artifact/annotation/ALL_SOURCES_ALL_FREQUENCIES_phenotype_to_genes.txt .
Where did you get your dataset from, in contrast? Is it definitely newer and better than what we download from the given URL?

I'm asking because although your new dataset (on /cluster/data/genomes/human/grch37/medgendb/latest/hpo/phenotype_to_gene.map ) has more genes in it (8000 vs 4000) but fewer HPO annotations (6000 vs 7000). Should we perhaps union the old and new datasets instead of replacing the old one with the new one? Or does the old dataset simply contain more HPO annotations for each gene, because the HPO annotations are less specific?

Johannes

---

Hi Johannes,

1 Building The Dataset
The two READMEs (/cluster/prj/medical_genomics/kb/hpo/external_package/README.download_log & /cluster/prj/medical_genomics/kb/hpo/local_build/grch37/hp_annotations_monthly_build87_20150701_20150701/README) explain which datasets were downloaded and how the latest mappings were built.

The mappings we have are a combination of the mappings from http://compbio.charite.de/hudson/job/hpo.annotations.monthly/lastStableBuild/artifact/annotation/ (pull in human disorder related phenotypes from OMIM and Orphanet) and from http://compbio.charite.de/hudson/job/hpo.ontology.uberpheno/234/artifact/data/uberphenoOut/ (this source pulls in the cross species mappings from MGI and ZFIN).

I used the ALL_SOURCES_ALL_FREQUENCIES_genes_to_phenotype.txt rather than the ALL_SOURCES_ALL_FREQUENCIES_phenotype_to_genes.txt.

genes_to_phenotypes has the primary mappings and phenotype_to_genes has the inferred mappings
(e.g. of inferred mappings - see the following relations http://www.human-phenotype-ontology.org/hpoweb/showterm?id=HP:0000234&GETIMAGE=ANCESTORS)
If a gene causes the phenotype "Abnormality of the head", then genes_to_phenotype will only contain this. While phenotype_to_genes associates the gene with "Abnormality of the head", "Abnormality of head or neck" and "Phenotypic abnormality" due to the relations that an "Abnormality of the head" is_a "Abnormality of head or neck" and so on.

This inferred mapping property is what causes the phenotype_to_genes file to have a lot more HPO annotations, but this can be recovered from the genes_to_phenotypes file.

Also I built the latest set on 20150701 which is the date of the latest build on the Charite website so you should get identical results if you tried to rerun all my instructions.

2 Inferred mappings or not
You guys probably do want the inferred mappings since for the most part inferring the relations up the DAG defined by the ontology is valid. Your call on rebuilding the entire set using their phenotype_to_genes files.
I think the easier thing would be to use my base mappings file and then run the canonicalization scripts from building GREAT ontologies (http://dev.stanford.edu/confluence/display/main/GREAT+Ontologies+HowTo#GREATOntologiesHowTo-Validateandcanonicalizeontology)

The script you want is $GITROOT/prj/greatOntologies/src/py/canonicalizeOntology.py. Make sure you use --redundantThresh=1.1 (default) so that you don't remove nodes (only GREAT needs to do that).

If you learn anything new about the HPO ontology during the process do note it at http://dev.stanford.edu/confluence/pages/viewpage.action?pageId=527199 so that it will be helpful the next time the ontology is updated.

Harendra

---

 Alex,

it was my impression that we had our own "walking up the HPO DAG" framework. Do you know if that's true? I only overheard some conversations with Gill and Morgan some time ago. If that's the case, maybe we don't need to do anything to canonicalize Harendra's file further.

On the other hand, it might be better in the long run to use the existing tools that also power GREAT in the background, because they're probably more reliable.

Johannes

---

Hey Johannes,

We do have a simple framework for traversing the DAG (see the DAG class) and do some of this canonicalization in several places e.g. here for Mesh supervision

For what it's worth, I think as long as what we do is transparent (even if this is just a reference to the GREAT tools), clean and centralized should be fine.  I'm putting an issue up around this last point of 'centralization'- we do have a very simple way of canonicalizing (ref'd above) and this should be gathered and put in the data_util file, if we continue using it

Thoughts?

Whenever we resolve this, let's also copy-paste the relevant bits of this thread to the issues comments so it's saved for reference (I know, I'm a broken record, but I'm in love...)

-Alex

---

FYI, some of the canonicalization stuff might change a bit for phenos when we switch to UMLS- all the more reason to start centralizing functions so this switch will be easier

I'll assign this to myself but whoever does this cleanup first nbd

---

Hey,

I don't really get what you mean in the second paragraph.

I looked at the GREAT tool that Harendra mentioned and it requires 5
input files whose format is only implicitly described as the output of
an array of further tools. The first issue is that I don't really
understand what these files actually contain, the second second I think
that I don't have the required information, and third I think it would
be overkill to get half of the GREAT backend running in order to add a
little more distant supervision. The last point in particular would take
a nontrivial amount of time.

The GREAT script apparently requires so much input because it is
concerned with removing as much redundant information as possible, which
we do not care about at all. Therefore I just started to write my own
"canonicalization": for each gene-phenotype association, add all
gene-(parent phenotype) associations until we hit the parent node
"phenotypic abnormality (118)". What I wanted to ask you is if there was
already code to do this "walk up the phenotype DAG", or if I was
reinventing the wheel even within our own codebase.

This does not appear to be the case. Yesterday, I wrote a few lines for
"walking up the DAG", which currently hangs somewhere in the
DAG.is_child method.

I agree about adding this conversation to the issue tracker, but I was
pretty sure that Aaron and Harendra don't actually read the load of
emails coming from our issue tracker (if they haven't turned the
notifications off already).

Johannes

---

The GREAT "ontology canonicalization" process is documented on Gill's lab wiki.  As Johannes mentions, the GREAT ontology canonicalization has two main steps: annotation closure and redundant term removal.  The annotation closure is what we want here: if gene g is annotated with term t, infer that g is annotated with all ancestors of t to the root of the DAG.  The redundant term removal is not relevant for us.  It's purpose in GREAT is to reduce the number of terms in an ontology so as to reduce the number of hypothesis tests, and thus the multiple hypothesis test correction factor.  If term t is annotated to the set of genes G, and parent(t) is annotated to the same exact set of genes G, then testing for significance of t is the same as testing for parent(t).  GREAT drops such terms "parent(t)" and keeps only the more specific term t.

I recommend that we implement the closure operation and apply it anytime that an ontology is loaded.  It's an idempotent operation; so, it's safe to apply multiple times, and applying it upon reading an ontology will make us robust to whether the version being read is already an annotation closure.

I'm not sure it will help, but here's Harendra and my JavaScript implementation of the annotation closure, which we use in our exome analysis system:

/*
    "_" here is the lodash JavaScript library
    hpoDag.bottomUp is a map from an HPO term to its parents in the DAG
    hpoAnnotations.geneToTerms is a map from an ENSEMBL gene ID to an array of the HPO terms with which it is annotated
*/
// Find the "closure" of a term or a list of terms, which includes all terms
// in the list and their ancestors up the DAG.
var termClosure = function(t) {
    var termList = _.isArray(t) ? _.uniq(t) : [t];
    var closure = {};

```
// Add a term and its ancestors to the closure.
var termClosureHelper = function(term) {
    if (!closure[term]) { // avoid retraversing a subtree
        closure[term] = true; // add the term
        _(hpoDag.bottomUp[term]).forEach(termClosureHelper).commit(); // add its ancestors
    }
};

    _(termList).forEach(termClosureHelper).commit();
    return _(closure).keys().value();
};
```

// Take the "closure" over HPO annotations.  If a gene is annotated with
// term X, also annotate it with all ancestors of X.
hpoAnnotations.geneToTerms = _(hpoAnnotations.geneToTerms).mapValues(function(terms) {
    return termClosure(terms);
}).value();

**Comment by Colossus** [comment]
@ajratner BTW: For the new genepheno data, please execute the following in onto:

```
./canonicalize_gene_phenotype.py | sort | uniq > data/canon_phenotype_to_ensgene.map
join -1 1 -2 1 \ 
  <(cat data/ensembl_genes.tsv | 
    grep -P '\tCANONICAL' | 
    awk -F'[:\t]' '{print $1, $2}' | 
    sort | uniq) \
  <(cat data/canon_phenotype_to_ensgene.map | 
    awk '{print $2, $1}' |
    sort | uniq) |
  awk '{print $3"\t"$2}' | sort > data/canon_phenotype_to_gene.map
```

(which is the now last block in make_all.sh)
